File: ra-idp/adapters/persistence/memory/login-attempt-throttle.ts
/**
 * Layer 4 — Adapter Layer (in-memory LoginAttemptThrottle)
 *
 * fixed-window counter + lockout キーの 2 段。test / memory モード用。
 * 本番は Redis adapter (../redis/login-attempt-throttle.ts) を使う。詳細は ADR-029。
 */

import type {
  LoginAttemptThrottle,
  LoginThrottleAcquireResult,
  LoginThrottleKind,
} from '../../../src/authentication/ports/login-attempt-throttle'

export interface LoginThrottleConfig {
  maxFailures: number
  windowSeconds: number
  lockoutSeconds: number
}

export interface LoginThrottleConfigs {
  account: LoginThrottleConfig
  ip: LoginThrottleConfig
}

interface CounterEntry {
  failures: number
  windowExpiresAtMs: number
}

interface LockEntry {
  expiresAtMs: number
}

export class InMemoryLoginAttemptThrottle implements LoginAttemptThrottle {
  private readonly counters = new Map<string, CounterEntry>()
  private readonly locks = new Map<string, LockEntry>()

  constructor(private readonly configs: LoginThrottleConfigs) {}

  async tryAcquire(
    kind: LoginThrottleKind,
    key: string,
    now: Date,
  ): Promise<LoginThrottleAcquireResult> {
    const lock = this.locks.get(lockKey(kind, key))
    if (!lock) return { allowed: true }
    const remainingMs = lock.expiresAtMs - now.getTime()
    if (remainingMs <= 0) {
      this.locks.delete(lockKey(kind, key))
      return { allowed: true }
    }
    return { allowed: false, retryAfterSeconds: Math.ceil(remainingMs / 1000) }
  }

  async recordFailure(
    kind: LoginThrottleKind,
    key: string,
    now: Date,
  ): Promise<{ locked: boolean; retryAfterSeconds?: number }> {
    const config = this.configs[kind]
    const counterK = counterKey(kind, key)
    const lockK = lockKey(kind, key)
    const nowMs = now.getTime()

    const existing = this.counters.get(counterK)
    let entry: CounterEntry
    if (!existing || existing.windowExpiresAtMs <= nowMs) {
      entry = { failures: 1, windowExpiresAtMs: nowMs + config.windowSeconds * 1000 }
    } else {
      entry = { failures: existing.failures + 1, windowExpiresAtMs: existing.windowExpiresAtMs }
    }
    this.counters.set(counterK, entry)

    if (entry.failures >= config.maxFailures) {
      this.counters.delete(counterK)
      this.locks.set(lockK, { expiresAtMs: nowMs + config.lockoutSeconds * 1000 })
      return { locked: true, retryAfterSeconds: config.lockoutSeconds }
    }
    return { locked: false }
  }

  async recordSuccess(kind: LoginThrottleKind, key: string): Promise<void> {
    this.counters.delete(counterKey(kind, key))
    this.locks.delete(lockKey(kind, key))
  }
}

function counterKey(kind: LoginThrottleKind, key: string): string {
  return `c:${kind}:${key}`
}

function lockKey(kind: LoginThrottleKind, key: string): string {
  return `l:${kind}:${key}`
}

File: ra-idp-go/internal/adapters/persistence/memory/login_attempt_throttle.go
package memory

import (
	"context"
	"math"
	"sync"
	"time"

	authports "ra-idp-go/internal/authentication/ports"
)

type LoginThrottleConfig struct {
	MaxFailures    int
	WindowSeconds  int
	LockoutSeconds int
}

type LoginThrottleConfigs struct {
	Account LoginThrottleConfig
	IP      LoginThrottleConfig
}

type loginCounter struct {
	failures  int
	expiresAt time.Time
}

type LoginAttemptThrottle struct {
	mu       sync.Mutex
	configs  LoginThrottleConfigs
	counters map[string]loginCounter
	locks    map[string]time.Time
}

func NewLoginAttemptThrottle(configs LoginThrottleConfigs) *LoginAttemptThrottle {
	return &LoginAttemptThrottle{
		configs: configs, counters: map[string]loginCounter{}, locks: map[string]time.Time{},
	}
}

func (t *LoginAttemptThrottle) TryAcquire(
	_ context.Context,
	kind authports.LoginThrottleKind,
	key string,
	now time.Time,
) (authports.LoginThrottleResult, error) {
	t.mu.Lock()
	defer t.mu.Unlock()
	lockKey := throttleKey(kind, key)
	expiresAt, ok := t.locks[lockKey]
	if !ok {
		return authports.LoginThrottleResult{Allowed: true}, nil
	}
	remaining := expiresAt.Sub(now)
	if remaining <= 0 {
		delete(t.locks, lockKey)
		return authports.LoginThrottleResult{Allowed: true}, nil
	}
	return authports.LoginThrottleResult{
		Allowed: false, Locked: true,
		RetryAfterSeconds: int(math.Ceil(remaining.Seconds())),
	}, nil
}

func (t *LoginAttemptThrottle) RecordFailure(
	_ context.Context,
	kind authports.LoginThrottleKind,
	key string,
	now time.Time,
) (authports.LoginThrottleResult, error) {
	t.mu.Lock()
	defer t.mu.Unlock()
	config := t.config(kind)
	counterKey := throttleKey(kind, key)
	counter, ok := t.counters[counterKey]
	if !ok || !now.Before(counter.expiresAt) {
		counter = loginCounter{failures: 1, expiresAt: now.Add(time.Duration(config.WindowSeconds) * time.Second)}
	} else {
		counter.failures++
	}
	t.counters[counterKey] = counter
	if counter.failures < config.MaxFailures {
		return authports.LoginThrottleResult{Allowed: true}, nil
	}
	delete(t.counters, counterKey)
	t.locks[counterKey] = now.Add(time.Duration(config.LockoutSeconds) * time.Second)
	return authports.LoginThrottleResult{
		Allowed: false, Locked: true, RetryAfterSeconds: config.LockoutSeconds,
	}, nil
}

func (t *LoginAttemptThrottle) RecordSuccess(
	_ context.Context,
	kind authports.LoginThrottleKind,
	key string,
) error {
	t.mu.Lock()
	defer t.mu.Unlock()
	key = throttleKey(kind, key)
	delete(t.counters, key)
	delete(t.locks, key)
	return nil
}

func (t *LoginAttemptThrottle) config(kind authports.LoginThrottleKind) LoginThrottleConfig {
	if kind == authports.LoginThrottleIP {
		return t.configs.IP
	}
	return t.configs.Account
}

func throttleKey(kind authports.LoginThrottleKind, key string) string {
	return string(kind) + ":" + key
}
